src/freenet/client/async/SplitFileInserterStorage.java

```diff
@@ -1760,4 +1762,16 @@ public void clearCooldown() {
         this.callback.clearCooldown();
     }
 
+    /** @return -1 if the insert has finished, 0 if has blocks to send, otherwise Long.MAX_VALUE. */
+    public long getWakeupTime(ClientContext context, long now) {
+        // LOCKING: hasFinished() uses (this), separate from cooldownLock.
+        // It is safe to use both here (on the request selection thread), one after the other.
+        if (hasFinished()) 
+            return -1;
+        if (noBlocksToSend())
+            return Long.MAX_VALUE;
```

Comment:
Thanks for the explanation on the expected usage of this code path. In that case, all should be fine.